File: services/scheduler.js
import { query } from '../config/db.js';
import { suspendPteroServer } from './pyrodactyl.js';
import { createNotification } from './notification.js';

async function suspendExpiredServers() {
  try {
    const expired = await query(
      "SELECT * FROM server_meta WHERE expires_at <= NOW() AND status = 'active'"
    );
    for (const row of expired) {
      try {
        await suspendPteroServer(row.ptero_server_id);
        await query("UPDATE server_meta SET status = 'suspended' WHERE id = ?", [row.id]);
        await createNotification(row.user_id, 'Server Expired', `Your server #${row.ptero_server_id} has been suspended due to expiry. Renew it to reactivate.`, 'warning', `/servers`);
        console.log(`Suspended server ${row.ptero_server_id} (expired)`);
      } catch (err) {
        console.error(`Failed to suspend server ${row.ptero_server_id}:`, err.message);
      }
    }
    if (expired.length > 0) {
      console.log(`Suspended ${expired.length} expired server(s)`);
    }
  } catch (err) {
    console.error('Scheduler check error:', err.message);
  }
}

function msUntilMidnight() {
  const now = new Date();
  const midnight = new Date(now);
  midnight.setDate(midnight.getDate() + 1);
  midnight.setHours(0, 0, 0, 0);
  return midnight.getTime() - now.getTime();
}

let schedulerTimer = null;
let schedulerRunning = false;

export function stopScheduler() {
  schedulerRunning = false;
  if (schedulerTimer) {
    clearTimeout(schedulerTimer);
    schedulerTimer = null;
  }
}

export function startScheduler() {
  if (schedulerRunning) return;
  schedulerRunning = true;
  suspendExpiredServers().catch(err => console.error('Initial scheduler run failed:', err.message));
  const tick = () => {
    if (!schedulerRunning) return;
    schedulerTimer = setTimeout(async () => {
      if (!schedulerRunning) return;
      try {
        await suspendExpiredServers();
      } catch (err) {
        console.error('Scheduled suspension failed:', err.message);
      }
      tick();
    }, msUntilMidnight());
  };
  tick();
  console.log('Server lifetime scheduler started');

  cleanupOldNotifications().catch(err => console.error('Initial notification cleanup failed:', err.message));
  cleanupOldActivityLogs().catch(err => console.error('Initial activity log cleanup failed:', err.message));
}

async function cleanupOldNotifications() {
  try {
    const result = await query(
      "DELETE FROM notifications WHERE created_at < NOW() - INTERVAL 90 DAY AND read = 1"
    );
    if (result.affectedRows > 0) {
      console.log(`Cleaned up ${result.affectedRows} old read notification(s)`);
    }
  } catch (err) {
    console.error('Notification cleanup error:', err.message);
  }
}

async function cleanupOldActivityLogs() {
  try {
    const result = await query(
      "DELETE FROM activity_logs WHERE created_at < NOW() - INTERVAL 90 DAY"
    );
    if (result.affectedRows > 0) {
      console.log(`Cleaned up ${result.affectedRows} old activity log(s)`);
    }
  } catch (err) {
    console.error('Activity log cleanup error:', err.message);
  }
}

File: internal/services/scheduler.go
package services

import (
	"fmt"
	"log"
	"sync"
	"time"

	"zerohost/dashboard/internal/database"
)

var (
	schedulerRunning bool
	schedulerMu      sync.Mutex
	stopChan         chan struct{}
)

func StartScheduler() {
	schedulerMu.Lock()
	if schedulerRunning {
		schedulerMu.Unlock()
		return
	}
	schedulerRunning = true
	stopChan = make(chan struct{})
	schedulerMu.Unlock()

	go func() {
		log.Println("Server lifetime scheduler started")
		suspendExpiredServers()

		for {
			next := msUntilMidnight()
			select {
			case <-time.After(next):
				suspendExpiredServers()
				cleanupOldNotifications()
				cleanupOldActivityLogs()
			case <-stopChan:
				return
			}
		}
	}()
}

func StopScheduler() {
	schedulerMu.Lock()
	defer schedulerMu.Unlock()
	if schedulerRunning {
		schedulerRunning = false
		close(stopChan)
	}
}

func msUntilMidnight() time.Duration {
	now := time.Now()
	midnight := time.Date(now.Year(), now.Month(), now.Day()+1, 0, 0, 0, 0, now.Location())
	return midnight.Sub(now)
}

func suspendExpiredServers() {
	rows, err := database.DB.Query(
		"SELECT * FROM server_meta WHERE expires_at <= NOW() AND status = 'active'",
	)
	if err != nil {
		log.Printf("Scheduler check error: %v", err)
		return
	}
	defer rows.Close()

	var expired []map[string]interface{}
	cols, _ := rows.Columns()
	for rows.Next() {
		vals := make([]interface{}, len(cols))
		valPtrs := make([]interface{}, len(cols))
		for i := range vals {
			valPtrs[i] = &vals[i]
		}
		rows.Scan(valPtrs...)
		row := make(map[string]interface{})
		for i, col := range cols {
			row[col] = vals[i]
		}
		expired = append(expired, row)
	}

	for _, row := range expired {
		pteroID := toInt64(row["ptero_server_id"])
		id := toInt64(row["id"])
		userID := toInt64(row["user_id"])

		if err := SuspendPteroServer(pteroID); err != nil {
			log.Printf("Failed to suspend server %d: %v", pteroID, err)
			continue
		}

		database.DB.Exec("UPDATE server_meta SET status = 'suspended' WHERE id = ?", id)
		CreateNotification(userID, "Server Expired",
			fmt.Sprintf("Your server #%d has been suspended due to expiry. Renew it to reactivate.", pteroID),
			"warning", nil)
		log.Printf("Suspended server %d (expired)", pteroID)
	}

	if len(expired) > 0 {
		log.Printf("Suspended %d expired server(s)", len(expired))
	}
}

func cleanupOldNotifications() {
	res, err := database.DB.Exec(
		"DELETE FROM notifications WHERE created_at < NOW() - INTERVAL 90 DAY AND is_read = 1",
	)
	if err != nil {
		log.Printf("Notification cleanup error: %v", err)
		return
	}
	if n, _ := res.RowsAffected(); n > 0 {
		log.Printf("Cleaned up %d old read notification(s)", n)
	}
}

func cleanupOldActivityLogs() {
	res, err := database.DB.Exec(
		"DELETE FROM activity_log WHERE created_at < NOW() - INTERVAL 90 DAY",
	)
	if err != nil {
		log.Printf("Activity log cleanup error: %v", err)
		return
	}
	if n, _ := res.RowsAffected(); n > 0 {
		log.Printf("Cleaned up %d old activity log(s)", n)
	}
}

func toInt64(v interface{}) int64 {
	if v == nil {
		return 0
	}
	switch val := v.(type) {
	case int64:
		return val
	case float64:
		return int64(val)
	case []byte:
		n := int64(0)
		for _, b := range val {
			if b >= '0' && b <= '9' {
				n = n*10 + int64(b-'0')
			} else {
				break
			}
		}
		return n
	}
	return 0
}
